client/state/data-layer/third-party/refer/index.js

```diff
@@ -38,12 +55,55 @@ const trackAffiliatePageLoad = action => {
 	);
 };
 
+const onTrackAffiliatePageLoadSuccess = ( action, response ) => {
+	if (
+		'object' !== typeof response ||
+		'object' !== typeof response.body ||
+		'object' !== typeof response.body.data
+	) {
+		debug( 'Unexpected referral response: ', response );
+		return; // Not possible.
+	}
+
+	const responseBody = response.body;
+	const responseData = responseBody.data;
+	const eventProps = pick( responseData, whitelistedEventProps );
+
+	eventProps.status = response.status || '';
+	eventProps.success = responseBody.success || '';
+	eventProps.description = responseBody.message || 'success';
+
+	analytics.tracks.recordEvent( 'calypso_refer_visit_response', eventProps );
+};
+
+const onTrackAffiliatePageLoadError = ( action, error ) => {
+	if (
+		'object' !== typeof error ||
+		'object' !== typeof error.response ||
+		'string' !== typeof error.response.text
+	) {
+		debug( 'Unexpected referral error: ', error );
+		return; // Not possible.
+	}
+
+	const response = error.response;
+	let responseBody = JSON.parse( response.text );
+	responseBody = 'object' === typeof responseBody ? responseBody : {};
+	const eventProps = pick( responseBody, whitelistedEventProps );
+
+	eventProps.status = response.status || '';
+	eventProps.success = responseBody.success || '';
+	eventProps.description = responseBody.message || 'error';
+
+	analytics.tracks.recordEvent( 'calypso_refer_visit_response', eventProps );
```

Comment:
AFAIK we don't want to use these methods directly as here, but [more like here](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/inline-help/index.jsx#L204). The latter is in a React component within a call to `connect` though, so I'm not sure how that would translate to this data layer module. 🤔 Just thinking out loud here.

Reply:
@jsnajdr or @dmsnell Can you help us answer this question? I want to record a Tracks event on HTTP success or failure. Have I done it properly here in the data-layer? Or is there a better way?

Reply:
Both handlers (success and error) should return a Redux action to be dispatched as a reaction to the response. And we have a dedicated Redux analytics library:
```js
import { recordTracksEvent } from 'state/analytics/actions';

function onTrackAffiliatePageLoadError( ... ) {
  /* ... */
  return recordTracksEvent( 'calypso_refer_visit_response', eventProps );
}
```

Reply:
Ah. Thank you so much!